### PR TITLE
libteec: Update TEEC_OpenSession() to pass connection data based on method

### PR DESCRIPTION
**Pull request by vesajaaskelainen**
TEE Client API specification states that only with login methods
TEEC_LOGIN_GROUP and TEEC_LOGIN_GROUP_APPLICATION connection data is
non-NULL.

In current Linux kernel API interfece it is not possible to handle custom
connection data so every login method needs to be manually passed thru
clnt_uuid field and handled specially both in user space and in kernel to
allow kernel to do proper authorization checking when opening an session.

This is related to: https://github.com/OP-TEE/optee_os/issues/3642

### CODE REVIEW

File: libteec/src/tee_client_api.c [hunk 1]
```diff
@@ -533,6 +531,52 @@ TEEC_Result TEEC_OpenSession(TEEC_Context *ctx, TEEC_Session *session,
 	uuid_to_octets(arg->uuid, destination);
 	arg->clnt_login = connection_method;
 
+	switch (connection_method) {
```
Comment (jenswi-linaro) [inline libteec/src/tee_client_api.c]: The documentation is nice, but it makes you lose the overview of this function.
Please move this into a helper function.

**Comment by jenswi-linaro** [comment]
> In current Linux kernel API interface it is not possible to handle custom
connection data so every login method needs to be manually passed thru
clnt_uuid field and handled specially both in user space and in kernel to
allow kernel to do proper authorization checking when opening an session.

I'd argue that it is possible since this is what you're doing in this patch. :-)
It's even done automatically by `TEEC_OpenSession()` with this.

File: libteec/src/tee_client_api.c (hunk 1 above)
Comment (vesajaaskelainen) [reply libteec/src/tee_client_api.c]: Ok.

**Comment by vesajaaskelainen** [comment]
> > In current Linux kernel API interface it is not possible to handle custom
> > connection data so every login method needs to be manually passed thru
> > clnt_uuid field and handled specially both in user space and in kernel to
> > allow kernel to do proper authorization checking when opening an session.
> 
> I'd argue that it is possible since this is what you're doing in this patch. :-)
> It's even done automatically by `TEEC_OpenSession()` with this.

I would say it is hackish. It does not allow login extensions that could provide free string as input.

**Comment by jenswi-linaro** [comment]
Whatever is passed needs to fit into a `TEE_Identity` when retrieved by the TA.

File: libteec/src/tee_client_api.c (hunk 1 above)
Comment (vesajaaskelainen) [reply libteec/src/tee_client_api.c:500]: moved and updated the PR

**Comment by vesajaaskelainen** [comment]
Just rebased this PR to latest master.

**Comment by jforissier** [comment]
Looks like this PR is mostly OK, @jenswi-linaro are you requesting some editing of the commit message? Speaking of which, there is a typo: `s/interfece/interface`.

**Comment by jenswi-linaro** [comment]
I'd like to have the commit message to describe what's added. The last paragraph doesn't add anything useful.

**Comment by vesajaaskelainen** [comment]
> I'd like to have the commit message to describe what's added. The last paragraph doesn't add anything useful.

Ok. Perhaps:
```
    libteec: Update TEEC_OpenSession() to pass connection data based on method
    
    Adds handling and descriptions for login methods TEE_IOCTL_LOGIN_PUBLIC,
    TEE_IOCTL_LOGIN_USER, TEE_IOCTL_LOGIN_GROUP, TEE_IOCTL_LOGIN_APPLICATION,
    TEE_IOCTL_LOGIN_USER_APPLICATION and TEE_IOCTL_LOGIN_GROUP_APPLICATION.
    
    TEE Client API specification states that only with login methods
    TEEC_LOGIN_GROUP and TEEC_LOGIN_GROUP_APPLICATION connection data is
    non-NULL and in these cases connection data is group identifier (gid_t).
```

**Comment by jenswi-linaro** [comment]
Sounds good, with that:
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by vesajaaskelainen** [comment]
> Sounds good, with that:
> `Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

Thanks for the review!

Updated the PR with those.

**Comment by jforissier** [comment]
Thanks @vesajaaskelainen.

I don't know what's wrong with checkpatch (Travis CI), it seems to be confused with the non-ASCII characters in your name perhaps. I checked locally, all is fine, and I could run our `checkpatch.sh` with no error.

**Pull request merged**